### PR TITLE
Support getting release versions of a provider from the terraform registry

### PR DESCRIPTION
**Pull request by pdecat**
This PR adds support for getting release versions of a provider from the terraform registry.

List versions:

```
# tfupdate release list -h
Usage: tfupdate release list [options] <SOURCE>

Arguments
  SOURCE             A path of release data source.
                     Valid format depends on --source-type option.
                       - github or gitlab:
                         owner/repo
                         e.g. terraform-providers/terraform-provider-aws
                      - tfregistryModule
                         namespace/name/provider
                         e.g. terraform-aws-modules/vpc/aws
                      - tfregistryProvider
                         namespace/type
                         e.g. hashicorp/aws

Options:
  -s  --source-type  A type of release data source.
                     Valid values are
                       - github (default)
                       - gitlab
                       - tfregistryModule
                       - tfregistryProvider

  -n  --max-length   The maximum length of list.

# tfupdate release list -s tfregistryProvider hashicorp/aws
2.69.0
2.70.0
3.0.0
3.1.0
3.2.0
3.3.0
3.4.0
3.5.0
3.6.0
3.7.0

# tfupdate release list -s tfregistryProvider datadog/datadog
2.12.1
2.13.0
```

Latest version:

```
# tfupdate release latest -h
Usage: tfupdate release latest [options] <SOURCE>

Arguments
  SOURCE             A path of release data source.
                     Valid format depends on --source-type option.
                       - github or gitlab:
                         owner/repo
                         e.g. terraform-providers/terraform-provider-aws
                      - tfregistryModule
                         namespace/name/provider
                         e.g. terraform-aws-modules/vpc/aws
                      - tfregistryProvider
                         namespace/type
                         e.g. hashicorp/aws

Options:
  -s  --source-type  A type of release data source.
                     Valid values are
                       - github (default)
                       - gitlab
                       - tfregistryModule
                       - tfregistryProvider

# tfupdate release latest -s tfregistryProvider hashicorp/aws
3.7.0

# tfupdate release latest -s tfregistryProvider datadog/datadog
2.13.0
```

### CODE REVIEW

File: command/release_list.go
Comment (minamijoyo) [inline command/release_list.go:75]: Could you add the same format example for the source argument as the release latest command?

File: tfregistry/provider.go
Comment (minamijoyo) [inline tfregistry/provider.go:45]: This link is for `:namespace/:type/versions`, not for `:namespace/:type`, and the later seems to be undocumented🤔 
I think it's a reasonable compromise here to use the undocumented API to simplify the implementation.
Could you remove the incorrect link and add a note that we are depending on the undocumented API which may be unstable to simplify the implementation?

Comment (minamijoyo) [reply tfregistry/provider.go:45]: The same link is copied to some places, so don't forget to fix all.

File: release/tfregistry_test.go
Comment (minamijoyo) [inline release/tfregistry_test.go:255]: Use `providerRes` instead of` moduleRes` to clarify the intention of the test case although the default value is nil.

File: release/tfregistry_test.go
Comment (minamijoyo) [inline release/tfregistry_test.go:390]: Use `providerRes` instead of `moduleRes`

File: tfregistry/provider_test.go
Comment (minamijoyo) [inline tfregistry/provider_test.go:28]: Why is the slice of `versions` unsorted? I think we implicitly expect the sorted list because the code comment in the TFRegistryProviderRelease.List() function says that versions are already in asc order unlike github.

File: tfregistry/provider_test.go
Comment (minamijoyo) [inline tfregistry/provider_test.go:57]: `invalid request (Type)`

File: tfregistry/provider_test.go
Comment (minamijoyo) [inline tfregistry/provider_test.go:68]: `invalid request (Type)`

File: tfregistry/provider_test.go
Comment (minamijoyo) [inline tfregistry/provider_test.go:57]: Why do we need this case? It seems to be valid dummy data (of course it's an invalid in the real world)

**Comment by minamijoyo** [comment]
@pdecat Thank you for working on this! It's a great addition 👍 
The implementation looks almost ok and I left some trivial review comments and questions. Please check them.

File: tfregistry/provider.go
Comment (pdecat) [reply tfregistry/provider.go:45]: Good point. I just assumed that feature was modeled just like the equivalent latest module feature.
I've opened a topic on the Hashicorp community forum to request more information about that undocumented providers API endpoint: https://discuss.hashicorp.com/t/terraform-registry-api-documentation-for-extended-providers-endpoint-features/14738

In the mean time, what about:
```suggestion
// This relies on a currently undocumented providers API endpoint which behaves exactly like the equivalent documented modules API endpoint.
// https://www.terraform.io/docs/registry/api.html#latest-version-for-a-specific-module-provider
```

File: command/release_list.go
Comment (pdecat) [reply command/release_list.go:78]: Done.

File: tfregistry/provider_test.go
Comment (pdecat) [reply tfregistry/provider_test.go]: No reason for the unsorted list, fixed.

File: tfregistry/provider_test.go
Comment (pdecat) [reply tfregistry/provider_test.go]: I just did a poor copy/paste job from the module test case, removed.

File: tfregistry/provider_test.go
Comment (pdecat) [reply tfregistry/provider_test.go]: Fixed.

**Review by pdecat — COMMENTED** [review]
Hi @minamijoyo, thanks for the review!
I believe I've addressed most issues and made a proposition for the missing documentation.

File: tfregistry/provider_test.go
Comment (minamijoyo) [inline tfregistry/provider_test.go:28]: The actual API response is ASC order. So I think `["3.5.0", "3.6.0", "3.7.0"]` is better than DESC for test data.

File: tfregistry/provider.go
Comment (minamijoyo) [reply tfregistry/provider.go:45]: Thank you for opening the topic on forum.

Even if the undocumented provider latest API is removed in the future, we can calculate the latest version from the documented provider versions API and switch the implementation if we need. However, we probably need a bit complex logic such as sorting semantic versions and excluding non-stable alpha/beta/rc releases. So I don't want to implement it until we actually need it. 

The suggested change for the code comment looks good to me. Could you fix all links?

File: tfregistry/provider_test.go
Comment (minamijoyo) [inline tfregistry/provider_test.go:28]: 👍

File: tfregistry/provider.go
Comment (pdecat) [reply tfregistry/provider.go:45]: Hi @minamijoyo, [Martin suggested](https://discuss.hashicorp.com/t/terraform-registry-api-documentation-for-extended-providers-endpoint-features/14738/2?u=pdecat) that tfupdate should only use official API endpoints that would be guaranteed to be supported by all registries, not just `registry.terraform.io`. I know that host is currently hardcoded in tfupdate's code but what about allowing to target other hosts by optionally passing it in the provider definition, e.g.:

```
tfupdate release list -s tfregistryProvider registry.terraform.io/hashicorp/aws
```

We could probably reuse / reimplement the version selection logic from the terraform CLI directly: https://github.com/hashicorp/terraform/blob/v0.13.3/internal/getproviders/registry_client.go#L364-L400

WDYT?

Comment (minamijoyo) [reply tfregistry/provider.go:45]: I didn't aware the definition of the minimum requirement of Registry API. 🤔 
I checked commit history and found that the doc is added by Martin a few months ago before v0.13 release 😉 
https://github.com/hashicorp/terraform/pull/25181
The doc looks a contract between Terraform CLI and Registry as if the Terraform core team says "Don't break the contract for backward compatibility." So we can assume that it's stable.

Although I don't think a third-party registry implementation is widely used and I don't have any plan to support it for now, but it makes sense that the tfupdate can be extended to support all valid provider addresses including third-party registry for future extension.  So I agree to implement using the minimum requirement API.

As implementation, we should not directly imports a specific Terraform version because the tfupdate should support multiple Terraform versions by nature. The implementation of registry protocol in hashicorp/terraform seems to be a bit inconsistent between provider and module, but we can probably use https://github.com/apparentlymart/go-versions for sorting versions which is used by provider protocol implementation.

Comment (minamijoyo) [reply tfregistry/provider.go:45]: Since switching API requires almost rewriting the current implementation in tfregistry package, If you want to get this feature as soon as possible, I don't mind to merge the current implementation as an experimental feature.

Comment (pdecat) [reply tfregistry/provider.go:45]: > If you want to get this feature as soon as possible, I don't mind a perfect stability for a first step and I can merge the current implementation as an experimental feature.

Ok, let's do this for now.

File: release/tfregistry.go
Comment (minamijoyo) [inline release/tfregistry.go:20]: Looks like there are still some uncorrected links left.

Comment (pdecat) [reply release/tfregistry.go]: Oops, fixed.

File: release/tfregistry.go
Comment (pdecat) [reply release/tfregistry.go]: Oops, fixed.

**Comment by pdecat** [comment]
Last comments fixed. Squashed all commits.

**Review by minamijoyo — APPROVED** [review]
LGTM 👍

**Comment by minamijoyo** [comment]
@pdecat I've merge it and added a warning to clarify that it's experimental for now and cut a new release v0.4.2 🚀
Thank you for your awesome work 🎉

**Pull request merged**
